Compatibility with lvk

G-Sommani opened the pull request:
To be more compatible with LVK, I:
- Changed the unit in the multiorder maps from deg-2 to sr-1;
- The unit is stored now in `TUNIT2` in the header;
- Changed the column name from `PROBABILITY DENSITY [deg-2]` to `[PROBDENSITY]`.

Moreover, in `plot.py` there were problems bacause the new matplotlib doesn't support anymore the attribute `QuadContourSet.collections`. This affects only `create_plot` when `dozoom` is set to `True`. I tried by using `get_paths()` to fix the problem, but I am not totally sure it is really doing what it should. At least in general it is not anymore giving error. However, is the option `dozoom=True` still used? It seems to me only redundant because there is already `create_plot_zoomed`. Could we get rid of the option `dozoom=True`?

tianluyuan commented on skyreader/utils/handle_map_data.py:
```diff
@@ -400,8 +400,10 @@ def prepare_multiorder_map(
+                max_nside, degrees=False,
+            ),
+            uniq_array,
+            unit="sr-1"
```
Is this variable `unit` ever used?

G-Sommani replied:
The variable `unit` sets the field `TUNIT2` in the header of the fits file, which is what LVK use to specify the units (here they show how their header works https://gracedb.ligo.org/api/superevents/S241210fu/files/bayestar.html). This variable here would make our headers more similar to their headers.

jessiethw replied:
yes, and if users read these with Astropy, the `TUNIT` field is also included with the data type and can easily be used for unit conversions, if needed

tianluyuan replied:
Thanks! I see now that it's a kwarg to mhealpy, so that makes sense.

tianluyuan reviewed (commented):
I think it's fine to remove the old `dozoom` option in `create_plot`, unless there's a reason to preserve it for legacy purposes? Would it make sense to match units and labels for the regular, non-multiorder FITs file?

ric-evans commented on skyreader/utils/handle_map_data.py:
```diff
@@ -400,8 +400,10 @@ def prepare_multiorder_map(
+            unit="sr-1"
         )
-        column_names = [f"{column_names[0]} DENSITY [deg-2]"]
+        column_names = ["PROBDENSITY"]
```
If you want, you can be fancy: define an `enum.StrEnum` class for titles, then use class members where needed. Personally, I'd leave this as is, but it's worth pointing out this option.

ric-evans replied:
`enum.StrEnum` is only for 3.11+, so ignore this suggestion :smile:

G-Sommani commented:
> I think it's fine to remove the old `dozoom` option in `create_plot`, unless there's a reason to preserve it for legacy purposes? Would it make sense to match units and labels for the regular, non-multiorder FITs file?

I agree that if there is a reason to preserve it for legacy purposes we should keep it. But how can we verify if there is any reason for doing so?
Regarding matching units, @jessiethw, would switching the probability to a density of probability in the flattened probability maps create problems in the programs for the FRA? Would they need to change accordingly? What do you think in general about that?

G-Sommani commented:
> > I think it's fine to remove the old `dozoom` option in `create_plot`, unless there's a reason to preserve it for legacy purposes? Would it make sense to match units and labels for the regular, non-multiorder FITs file?
> 
> I agree that if there is a reason to preserve it for legacy purposes we should keep it. But how can we verify if there is any reason for doing so? Regarding matching units, @jessiethw, would switching the probability to a density of probability in the flattened probability maps create problems in the programs for the FRA? Would they need to change accordingly? What do you think in general about that?

I saw now the reference in [issue 27](https://github.com/icecube/skyreader/issues/27). Makes sense to do everything with healpy functions. Maybe better in a dedicated pull request?

jessiethw commented:
> > I think it's fine to remove the old `dozoom` option in `create_plot`, unless there's a reason to preserve it for legacy purposes? Would it make sense to match units and labels for the regular, non-multiorder FITs file?
> 
> I agree that if there is a reason to preserve it for legacy purposes we should keep it. But how can we verify if there is any reason for doing so? Regarding matching units, @jessiethw, would switching the probability to a density of probability in the flattened probability maps create problems in the programs for the FRA? Would they need to change accordingly? What do you think in general about that?

It would require a change so that we handle them in FRA as probability density rather than probability, but it should be a constant offset in each pixel in the case of a flattened map (corresponding to the pixel area) so it shouldn't be a large change there

tianluyuan commented:
> > I think it's fine to remove the old `dozoom` option in `create_plot`, unless there's a reason to preserve it for legacy purposes? Would it make sense to match units and labels for the regular, non-multiorder FITs file?
> 
> I agree that if there is a reason to preserve it for legacy purposes we should keep it. But how can we verify if there is any reason for doing so? 

Maybe sending a notice one of the realtime slack channels and if no one complains that would suffice? Also, if the functionality is changed/broken due to a change in the dependency then it doesn't make sense to preserve.

G-Sommani commented:
> > > I think it's fine to remove the old `dozoom` option in `create_plot`, unless there's a reason to preserve it for legacy purposes? Would it make sense to match units and labels for the regular, non-multiorder FITs file?
> > 
> > 
> > I agree that if there is a reason to preserve it for legacy purposes we should keep it. But how can we verify if there is any reason for doing so? Regarding matching units, @jessiethw, would switching the probability to a density of probability in the flattened probability maps create problems in the programs for the FRA? Would they need to change accordingly? What do you think in general about that?
> 
> It would require a change so that we handle them in FRA as probability density rather than probability, but it should be a constant offset in each pixel in the case of a flattened map (corresponding to the pixel area) so it shouldn't be a large change there

Thanks! I changed the units for the flattened map.

In the header, I also added the UTC date of the event, and I changed the name of the MJD and its comment, so that everything is as in the headers of LVK

tianluyuan commented:
Thanks @G-Sommani  for sending the message on slack about the old zoomed version. Should we go ahead and remove that?

I remember now that LVK also releases flattened maps. Is that something that we would do as well, i.e. release both multi-order and standard healpix? If so, perhaps it would be worth checking whether LVK flattened maps are in probability density or probability per pixel, and match to that. I sort of remember @jessiethw mentioning that those were per pixel now.

Once that's resolved should we try to get this merged in?

G-Sommani commented:
> Thanks @G-Sommani for sending the message on slack about the old zoomed version. Should we go ahead and remove that?
> 
> I remember now that LVK also releases flattened maps. Is that something that we would do as well, i.e. release both multi-order and standard healpix? If so, perhaps it would be worth checking whether LVK flattened maps are in probability density or probability per pixel, and match to that. I sort of remember @jessiethw mentioning that those were per pixel now. I'm not strongly tied to either one but just wanted to bring that up. Sorry for the added noise if we were to revert those changes!
> 
> Once that's resolved should we try to get this merged in?

I removed the option `dozoom` from `create_plot()`. In future it would be nice to have everything handled with healpy as suggested in [issue 27](https://github.com/icecube/skyreader/issues/27), but I guess better to dedicate a pull request just for that.

Regarding the LVK flattened maps, if I am not wrong, the aim would be to publish only multiorder maps, in case @blaufuss can confirm.

blaufuss commented:
From my point of view, muilti-order maps make for much smaller objects with high resolution, and are pretty standard  in use now adays thanks to LVK. 

But, that said, have no objection if we wanted to post both to a website archive, but would prefer to include the multi-order map as our default product...

jessiethw commented:
> Thanks @G-Sommani for sending the message on slack about the old zoomed version. Should we go ahead and remove that?
> 
> I remember now that LVK also releases flattened maps. Is that something that we would do as well, i.e. release both multi-order and standard healpix? If so, perhaps it would be worth checking whether LVK flattened maps are in probability density or probability per pixel, and match to that. I sort of remember @jessiethw mentioning that those were per pixel now. I'm not strongly tied to either one but just wanted to bring that up. Sorry for the added noise if we were to revert those changes!

sorry for the delay. Yes LVK does have both, but (same as Erik suggests for us) multiorder is the default for them. The flattened from LVK are per pixel, so we could stick with that for the flattened.

G-Sommani commented:
> > Thanks @G-Sommani for sending the message on slack about the old zoomed version. Should we go ahead and remove that?
> > I remember now that LVK also releases flattened maps. Is that something that we would do as well, i.e. release both multi-order and standard healpix? If so, perhaps it would be worth checking whether LVK flattened maps are in probability density or probability per pixel, and match to that. I sort of remember @jessiethw mentioning that those were per pixel now. I'm not strongly tied to either one but just wanted to bring that up. Sorry for the added noise if we were to revert those changes!
> 
> sorry for the delayed response from me. Yes LVK does have both, but (same as Erik suggests for us) multiorder is the default for them. The flattened from LVK are per pixel, so we could stick with that for the flattened.

Thanks, @Jessie! So better to remove the probability density from the flat maps? For me it is the same, just let me know which version would be the most preferable

jessiethw commented:
> > > Thanks @G-Sommani for sending the message on slack about the old zoomed version. Should we go ahead and remove that?
> > > I remember now that LVK also releases flattened maps. Is that something that we would do as well, i.e. release both multi-order and standard healpix? If so, perhaps it would be worth checking whether LVK flattened maps are in probability density or probability per pixel, and match to that. I sort of remember @jessiethw mentioning that those were per pixel now. I'm not strongly tied to either one but just wanted to bring that up. Sorry for the added noise if we were to revert those changes!
> > 
> > 
> > sorry for the delayed response from me. Yes LVK does have both, but (same as Erik suggests for us) multiorder is the default for them. The flattened from LVK are per pixel, so we could stick with that for the flattened.
> 
> Thanks, @jessie! So better to remove the probability density from the flat maps? For me it is the same, just let me know which version would be the most preferable

Ok, if it's all the same on your side let's stick with probability per pixel, to stay consistent with the same for LVK.

G-Sommani commented on skyreader/plot/plot.py:
```diff
@@ -42,7 +41,7 @@
     PLOT_SIZE_Y_IN: float = 3.85
     PLOT_SIZE_X_IN: float = 6
-    PLOT_DPI_STANDARD = 150
+    PLOT_DPI_STANDARD = 300
```
I noticed that for some small contours with the probability maps the small definition was a problem when creating the all-sky map. Just making the definition double seemed to fix the problem.

G-Sommani commented:
Flattened map is now again with probability per pixel. If this is OK, I would proceed with the merge (also considering the higher dpi for the all-sky map)

jessiethw commented on skyreader/utils/handle_map_data.py:
```diff
@@ -369,21 +369,20 @@ def prepare_flattened_map(
-        column_names = ["PROBDENSITY"]
-        column_units = ["sr-1"]
-    return equatorial_map, column_names, column_units
+        column_names = ["PROBABILITY"]
```
very minor comment: LVK just uses `PROB` as the column name for their flattened maps. also probably worth including units for flattened as well, they use `pix-1`

G-Sommani replied:
Done!

tianluyuan commented on skyreader/plot/plot.py:
```diff
@@ -193,116 +176,23 @@ def create_plot(
                 ra, dec, map_to_plot, levels=[level], colors=[color]
             )
-            cs_collections.append(contour_set.collections[0])
+            cs_collections.append(contour_set.get_paths()[0])
```
I thought this change wasn't needed anymore with the removal of `do_zoom` option. Does it work as is?

G-Sommani replied:
You are right, `cs_collections` and `leg_element` are not needed anymore. I now removed all the components which are unnecessary

tianluyuan commented on skyreader/plot/plot.py:
```diff
@@ -825,8 +704,9 @@ def bounding_box(ra, dec, theta, phi):
             equatorial_map,
             coord='C',
             column_names=column_names,
+            column_units=["pix-1"],
```
Maybe this unit should be set to a variable in the `if llh_map .. else` and then passed through here.

tianluyuan commented on skyreader/utils/handle_map_data.py:
```diff
@@ -400,8 +400,10 @@ def prepare_multiorder_map(
+            unit="sr-1"
         )
-        column_names = [f"{column_names[0]} DENSITY [deg-2]"]
+        column_names = ["PROBDENSITY"]
```
Does `column_names` also need to be set in the `llh_map` case?

G-Sommani replied:
In the `llh_map` case it just keeps the same list which was set by `prepare_flattened_map`

tianluyuan reviewed (commented):
There seems to be a few missing units but other than that it looks good. Thanks for the work!
